src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenAsyncIteratorTests.cs

```diff
@@ -631,15 +631,186 @@ await foreach (var s in M())
 ref struct S
 {
 }";
+
+            var expectedDiagnostics = new[]
+            {
+                // source(4,65): error CS0306: The type 'S' may not be used as a type argument
+                //     static async System.Collections.Generic.IAsyncEnumerable<S> M()
+                Diagnostic(ErrorCode.ERR_BadTypeArgument, "M").WithArguments("S").WithLocation(4, 65)
+            };
+
             var comp = CreateCompilationWithAsyncIterator(source, options: TestOptions.DebugExe);
+            comp.VerifyDiagnostics(expectedDiagnostics);
+
+            comp = CreateCompilationWithAsyncIterator(source, options: TestOptions.DebugExe, parseOptions: TestOptions.RegularNext);
+            comp.VerifyDiagnostics(expectedDiagnostics);
+
+            comp = CreateCompilationWithAsyncIterator(source, options: TestOptions.DebugExe, parseOptions: TestOptions.Regular12);
             comp.VerifyDiagnostics(
                 // source(4,65): error CS0306: The type 'S' may not be used as a type argument
                 //     static async System.Collections.Generic.IAsyncEnumerable<S> M()
                 Diagnostic(ErrorCode.ERR_BadTypeArgument, "M").WithArguments("S").WithLocation(4, 65),
-                // source(11,24): error CS4012: Parameters or locals of type 'S' cannot be declared in async methods or async lambda expressions.
+                // source(11,24): error CS8652: The feature 'Ref and unsafe in async and iterator methods' is currently in Preview and *unsupported*. To use Preview features, use the 'preview' language version.
                 //         await foreach (var s in M())
-                Diagnostic(ErrorCode.ERR_BadSpecialByRefLocal, "var").WithArguments("S").WithLocation(11, 24)
-                );
+                Diagnostic(ErrorCode.ERR_FeatureInPreview, "var").WithArguments("Ref and unsafe in async and iterator methods").WithLocation(11, 24));
+        }
+
+        [Fact]
+        public void RefStructElementType_NonGeneric()
+        {
+            string source = """
+                using System.Threading.Tasks;
+
+                class C
+                {
+                    public E GetAsyncEnumerator() => new E();
+                    static async Task Main()
+                    {
+                        await foreach (var s in new C())
+                        {
+                            System.Console.Write(s.F);
+                        }
+                    }
+                }
+                class E
+                {
+                    bool _done;
+                    public S Current => new S { F = 123 };
+                    public async Task<bool> MoveNextAsync()
+                    {
+                        await Task.CompletedTask;
```

Comment:
>CompletedTask [](http://example.com/codeflow?start=35&length=13)

For the purpose of the test, it is probably better to Yield if we are going to await.